numpy/lib/_polynomial_impl.py

```diff
@@ -760,7 +760,9 @@ def polyval(p, x):
 
     """
     p = NX.asarray(p)
-    if isinstance(x, poly1d):
+    if p.size == 0:
+        return NX.int64(0).astype(p.dtype)
```

Comment:
Having special paths for `size == 0` has a code smell, and not worth optimizing for (unless there is a good reason).

To me this looks wrong for two reasons:
1. It matches any shape, even high dimensional arrays, and flattens them.
2. More importantly, why is int64 even the correct result?

Am I missing a reason for why this is needed at all?

Reply:
The return type is not int64, as there is the conversion using `astype(p.dtype)`. But you are correct the handing of shapes is not correct. The following case fails:
```
v = polyval(np.array([], dtype=np.float32), [5.0, 2.0])
```
The argument `p` could be higher dimensional, but that gives the same results (there is no broadcasting)

@luxedo You could try to replace
```
    return NX.int64(0).astype(p.dtype)
```
with 
```
    return np.zeros_like(x).astype(p.dtype)
```
and add the example as a test case?

Reply:
Thanks, I'll have to think about it.  To some degree it would be nice if we could just do the identical "operation" in all cases.  Like using `(x * y + pv[:0])` to get the shape/dtype of the initial array and only then do the loop.

Reply:
Yes. I merged your changes and I will wrap it up if something comes up

Reply:
Two tests are failing:

1. `assert np.isscalar(v)` 

https://github.com/numpy/numpy/blob/d451e470dd90bdc5edb73137e2f4a71e9b06c936/numpy/lib/tests/test_polynomial.py#L323

This happens because `y` is converted to an array  here:
https://github.com/numpy/numpy/blob/d451e470dd90bdc5edb73137e2f4a71e9b06c936/numpy/lib/_polynomial_impl.py#L772

But the expected return would be a 0-d array

2.  `poly1d.__call__`

https://github.com/numpy/numpy/blob/d451e470dd90bdc5edb73137e2f4a71e9b06c936/numpy/lib/tests/test_polynomial.py#L59-L60
This gives two completely different arrays. I'm going to investigate further
```
E       AssertionError: 
E       Arrays are not equal
E       
E       (shapes (3,), (5,) mismatch)
E        ACTUAL: array([18., 11.,  6.])
E        DESIRED: array([ 9., 12., 16.,  8.,  6.])
```